src/mscorlib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -427,19 +471,21 @@ private bool TryInsert(TKey key, TValue value, InsertionBehavior behavior)
                 count++;
             }
 
-            entries[index].hashCode = hashCode;
-            entries[index].next = buckets[targetBucket];
-            entries[index].key = key;
-            entries[index].value = value;
+            ref Entry entry = ref entries[index];
+            entry.hashCode = hashCode;
+            entry.next = buckets[targetBucket];
+            entry.key = key;
+            entry.value = value;
             buckets[targetBucket] = index;
             version++;
 
             // If we hit the collision threshold we'll need to switch to the comparer which is using randomized string hashing
             // i.e. EqualityComparer<string>.Default.
 
-            if (collisionCount > HashHelpers.HashCollisionThreshold && comparer is NonRandomizedStringEqualityComparer)
+            if (typeof(TKey) == typeof(string) && collisionCount > HashHelpers.HashCollisionThreshold && comparer is NonRandomizedStringEqualityComparer)
```

Comment:
I do not think `typeof(TKey) == typeof(string)` check is going to help anything here.

Reply:
@jkotas If I am reading that properly you are killing the whole branch when ```TKey``` is not an string and struct. You are also adding a branching check when ```TKey``` is a class. Could be a win if and only if the check introduced doesnt make its performance worse.

Reply:
Right. Maybe the check can stay - but it should be after `collisionCount > HashHelpers.HashCollisionThreshold` to not regress the class path.

Reply:
Correct on that front.

Reply:
Change change it to a struct check?

Reply:
How would you do that in that case?

Reply:
```csharp
if (default(TKey) == null && collisionCount > HashHelpers.HashCollisionThreshold ...
```

Reply:
Nice trick :D